src/sql_types.rs

```diff
@@ -518,10 +528,107 @@ pub fn load_sql_context(_config: &Config) -> Result<Arc<Context>, String> {
+                mtype.details = match mtype.category {
+                    TypeCategory::Enum => context.enums.get(oid).cloned().map(TypeDetails::Enum),
+                    TypeCategory::Composite => context
+                        .composites
+                        .iter()
+                        .find(|c| c.oid == *oid)
+                        .cloned()
+                        .map(TypeDetails::Composite),
+                    TypeCategory::Table => context.tables.get(oid).cloned().map(TypeDetails::Table),
+                    TypeCategory::Array => {
+                        // We can't cross-reference with `context.types` here as it is already mutably borrowed,
+                        // so we instead memorize the reference to process later
+                        if let Some(element_oid) = mtype.array_element_type_oid {
+                            array_types.insert(*oid, element_oid);
+                        }
+                        None
+                    }
+                    _ => None,
+                };
+            }
+        }
+
+        // Ensure the types are ordered so that we don't run into a situation where we can't
+        // update the type anymore as it has been referenced but the type details weren't completed yet
+        let referenced_types = array_types.values().map(|k| *k).collect::<Vec<_>>();
+        let mut ordered_types = array_types
+            .iter()
+            .map(|(k, v)| (*k, *v))
+            .collect::<Vec<_>>();
+        // We sort them by their presence in referencing. If the type has been referenced,
+        // it should be at the top.
+        ordered_types.sort_by(|(k1, _), (k2, _)| {
+            if referenced_types.contains(k1) && referenced_types.contains(k2) {
+                Ordering::Equal
+            } else if referenced_types.contains(k1) {
+                Ordering::Less
+            } else {
+                Ordering::Greater
+            }
+        });
+
+        // Now we're ready to process array types
+        for (array_oid, element_oid) in ordered_types {
+            // SAFETY: unwrap() should be fine here as we just constructed this OID mapping
+            let element_t = context.types.remove(&element_oid).unwrap();
+            // We remove it from the map to ensure there is no mutability conflict
+            // SAFETY: this should work as this type is not referenced anywhere just yet
+            let array_t = context.types.get_mut(&array_oid).unwrap();
+            if let Some(array) = Arc::get_mut(array_t) {
+                // It should be possible to get a mutable reference to type at this point
+                // as there are no other references to this Arc at this point.
+                array.details = Some(TypeDetails::Element(element_t.clone()));
+            }
+            // Put the element type back
+            context.types.insert(element_oid, element_t);
+        }
+        context
```

Comment:
this all feels a bit unsafe

For example, are we 100% sure that no two array types can reference the same element type? If they do, this will panic. Updates to the sql context query could also break assumptions in some cases
```rust
let element_t = context.types.remove(&element_oid).unwrap();
```
It feels like edge cases are possible here and I'd be afraid to mess with the logic

Any errors would cause total failure of the API without indicating what went wrong

Can this be re-factored to work immutably and not require unwraps? If we have its slower and/or requires more unwrapping at runtime I'm okay with that trade because we can throw more informative error messages.

Since this will be cached we don't need to be as performance sensitive

Reply:
Happy to work through this to make sure it is safe.

The particular line is presumed safe because we return the element to the HashMap so that two types that refer to the same element type should work, and it will be available to other types in that loop.

The other lines are safe with unwrapping because we are populating that very HashMap from a known source.

However, for extra caution, we can make it use `if let Some` instead of unwrapping it so that it doesn't crash. In the `else` branch, we can issue a warning or even an error as it is unexpected behavior.

Reply:
I rewrote this part to be more on the defensive side, and it'll do proper reporting if something truly unexpected happens.